File: src/main/python/lab920_for_each_sink/streamRecordThroughForEachApp.py
"""
  Analyzes the records on the stream and send each record to a debugger class.

  @author rambabu.posa
"""
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, StringType, IntegerType
import logging

logging.debug("-> start")

# Creates a session on a local master
spark = SparkSession.builder.appName("Read lines over a file stream") \
    .master("local[*]").getOrCreate()

# The record structure must match the structure of your generated
# record
# (or your real record if you are not using generated records)
recordSchema = StructType([StructField('fname', StringType(), True),
                           StructField('mname', StringType(), True),
                           StructField('lname', StringType(), True),
                           StructField('age', IntegerType(), True),
                           StructField('ssn', StringType(), True)])

# Reading the record is always the same
df = spark.readStream.format("csv") \
    .schema(recordSchema) \
    .csv("/tmp/")
# use for windows
#   .csv("C:/tmp/")


class RecordLogDebugger:
    def open(self, partition_id, epoch_id):
        # Open connection. This method is optional in Python.
        True

    def process(self, row):
        # Write row to connection. This method is NOT optional in Python.
        logging.info('processing {}'.format(row))

    def close(self, error):
        # Close the connection. This method in optional in Python.
        pass


query = df.writeStream.outputMode("update") \
    .foreach(RecordLogDebugger()) \
    .start()

query.awaitTermination()

spark.stop()

logging.debug("<- end")

File: src/main/java/net/jgp/books/spark/ch10/lab920_for_each_sink/StreamRecordThroughForEachApp.java
package net.jgp.books.spark.ch10.lab920_for_each_sink;

import java.util.concurrent.TimeoutException;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.streaming.OutputMode;
import org.apache.spark.sql.streaming.StreamingQuery;
import org.apache.spark.sql.streaming.StreamingQueryException;
import org.apache.spark.sql.types.StructType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import net.jgp.books.spark.ch10.x.utils.streaming.lib.StreamingUtils;

/**
 * Analyzes the records on the stream and send each record to a debugger
 * class.
 * 
 * @author jgp
 *
 */
public class StreamRecordThroughForEachApp {
  private static Logger log =
      LoggerFactory.getLogger(StreamRecordThroughForEachApp.class);

  public static void main(String[] args) {
    StreamRecordThroughForEachApp app = new StreamRecordThroughForEachApp();
    try {
      app.start();
    } catch (TimeoutException e) {
      log.error("A timeout exception has occured: {}", e.getMessage());
    }
  }

  private void start() throws TimeoutException {
    log.debug("-> start()");

    SparkSession spark = SparkSession.builder()
        .appName("Read lines over a file stream")
        .master("local")
        .getOrCreate();

    StructType recordSchema = new StructType()
        .add("fname", "string")
        .add("mname", "string")
        .add("lname", "string")
        .add("age", "integer")
        .add("ssn", "string");

    Dataset<Row> df = spark
        .readStream()
        .format("csv")
        .schema(recordSchema)
        .csv(StreamingUtils.getInputDirectory());

    StreamingQuery query = df
        .writeStream()
        .outputMode(OutputMode.Update())
        .foreach(new RecordLogDebugger())
        .start();

    try {
      query.awaitTermination(60000);
    } catch (StreamingQueryException e) {
      log.error(
          "Exception while waiting for query to end {}.",
          e.getMessage(),
          e);
    }

    log.debug("<- start()");
  }
}
